db/migrate/20170914202111_add_material_type_to_external_activity.rb

```diff
@@ -0,0 +1,17 @@
+class AddMaterialTypeToExternalActivity < ActiveRecord::Migration
+
+  def change
+    add_column :external_activities, :material_type, :string
+
+    ExternalActivity.all.each do |external_activity|
+        if ! external_activity.template_type.nil? &&
+            ! external_activity.template_type.blank?
+            external_activity.update_attributes(:material_type => external_activity.template_type)
+        else
+            external_activity.update_attributes(:material_type => 'Activity')
+        end
+    end
```

Comment:
This method of iterating object is not efficient.  `ExternalActivity.all.each` will load every ExternalActivity into memory before iterating them.  This might not be a problem for ExternalActivities but in general it is good practice to use `ExternalActivity.find_each` instead.  By default find_each iterates through the instances in batches of 1000.  You can configure that lower if need be.

Another thing that we try to do (we are not always successful) is to try to make the migrations forward compatible. Despite the recommendations from the Rails documentation, we still run all of the migrations from the beginning of time.  

In this case, if the ExternalActivity model is removed or renamed, then this migration would stop working. There are at least two solutions to this.
1. Use SQL to make the changes.  This will keep the migration independent of any future model changes. It also allows you to be very efficient when make changes to large numbers of models.  However it does mean the migration could be more error prone and might be specific to the particular database we are using.
2. Re-define the Model class right in the migration.  A simple version of this is here: http://github.com/concord-consortium/rigse/blob/14a65077d4b3e04059bb9b59704255383bbca8e7/db/migrate/20160106034419_remove_is_member_from_admin_project_user.rb#L2
And here is an extreme example: http://github.com/concord-consortium/rigse/blob/14a65077d4b3e04059bb9b59704255383bbca8e7/db/migrate/20160629104824_add_teacher_info_to_report_learner.rb#L7 

In this specific case you could use a model definition like
```
  class ExternalActivity < ActiveRecord::Base
    self.table_name = 'external_activities'
  end
```

And more efficient queries would be:
```
ExternalActivity.where('template_type is null').update_all(material_type: 'Activity')
ExternalActivity.where(template_type: 'Activity').update_all(material_type: 'Activity')
ExternalActivity.where(template_type: 'Investigation').update_all(material_type: 'Investigation')
```